### ISSUE TITLE
Microsoft.AspNetCore.Authentication.Negotiate  OnAuthenticated Claims Transformation

### ISSUE THREAD

**Issue by treborgero** [dotnet/aspnetcore#21557]
### Describe the bug
Assembly: Microsoft.AspNetCore.Authentication.Negotiate v3.1.3

The "**OnAuthenticated**" and other event properties of the "**NegotiateEvents**" class are not being called when running the app with IIS. This prevents, for example, implementing proper claims transformation. Please note this behavior is not consistent and it is not documented anywhere.

After further investigation of the source code, this happens because the current logic is delegating the authentication back to IIS and skips the events.

### To Reproduce
First, enable windows authentication in IISExpress or IIS (via the web.config). Then, run the app with the configured events and they will not be called.
```
services.AddAuthentication()
                .AddNegotiate(o =>
                {
                    o.Events = new NegotiateEvents()
                    {
                        OnAuthenticated = authContext =>
                        {
                            //Never hits when using IIS
                            return Task.CompletedTask;
                        }
                    };
                });
```

### Further technical details
- ASP.NET Core version 3.1
- Microsoft.AspNetCore.Authentication.Negotiate v3.1.3
- Visual Studio Community 2019 Version 16.5.4

**Comment by treborgero** [comment]
@Tratcher

**Comment by Tratcher** [comment]
This is the intended behavior and it's covered in this doc:
https://docs.microsoft.com/en-us/aspnet/core/security/authentication/windowsauth?view=aspnetcore-3.1&tabs=visual-studio#kestrel
```
The Negotiate handler detects if the underlying server supports Windows Authentication natively and if it's enabled. If the server supports Windows Authentication but it's disabled, an error is thrown asking you to enable the server implementation. When Windows Authentication is enabled in the server, the Negotiate handler transparently forwards to it.
```

AddNegotiate is intended only for use with Kestrel. For consistent claims transformation across servers and auth implementations you should use IClaimsTransformation. @HaoK Is there a doc covering IClaimsTransformation? I'm not seeing one.

**Comment by HaoK** [comment]
Hrm I don't think we have official docs, guess we should probably add a section for that

**Comment by treborgero** [comment]
Thank you for the quick response. Hopefully it can be documented and prevent some time loss.

Though I'm curious... Your "[NegotiateHandler](https://github.com/dotnet/aspnetcore/blob/master/src/Security/Authentication/Negotiate/src/NegotiateHandler.cs)" could've potentially killed two birds with one stone and allowed developers the ability to implement "mixed" authentication a little more easier (e.g. in SPA scenarios with JWT, Negotiate, Certificates, etc.).  

In the past I created a handler leveraging .NET Core's "MS-ASPNETCORE-WINAUTHTOKEN" forwarding which allowed me to use Feature Delegation in the web.config:

Handler Example:
```
//The hammer handler
internal class WindowsAuthenticationHandler : AuthenticationHandler<WindowsAuthenticationOptions>
    {
        private const string MSAspNetCoreWinAuthToken = "MS-ASPNETCORE-WINAUTHTOKEN";


        public WindowsAuthenticationHandler(IOptionsMonitor<WindowsAuthenticationOptions> options, ILoggerFactory logger, UrlEncoder encoder, IDataProtectionProvider dataProtection, ISystemClock clock)
            : base(options, logger, encoder, clock)
        { }


        protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
        {
            var user = GetUser();

            if (user != null)
            {
                var claims = new List<Claim>();


                if (Options.OnCreateClaims != null)
                {
                    claims = await Options.OnCreateClaims(user);
                }
                else
                {
                    claims = new List<Claim>(user.Claims);
                }


                if (claims?.Count > 0)
                {
                    var userIdentity = new ClaimsIdentity(claims, Options.Challenge);
                    var userPrincipal = new ClaimsPrincipal(userIdentity);
                    var ticket = new AuthenticationTicket(userPrincipal, Options.Challenge);
                    return AuthenticateResult.Success(ticket);
                }
            }

            return AuthenticateResult.NoResult();
        }

        private WindowsPrincipal GetUser()
        {
            WindowsPrincipal user = null;

            var tokenHeader = Context.Request.Headers[MSAspNetCoreWinAuthToken];

            int hexHandle;
            if (!StringValues.IsNullOrEmpty(tokenHeader)
                && int.TryParse(tokenHeader, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out hexHandle))
            {
                // Always create the identity if the handle exists, we need to dispose it so it does not leak.
                var handle = new IntPtr(hexHandle);
                var winIdentity = new WindowsIdentity(handle);

                // WindowsIdentity just duplicated the handle so we need to close the original.
                NativeMethods.CloseHandle(handle);

                Context.Response.RegisterForDispose(winIdentity);

                user = new WindowsPrincipal(winIdentity);
            }

            return user;
        }

        protected override Task HandleChallengeAsync(AuthenticationProperties properties)
        {
            // We would normally set the www-authenticate header here, but IIS does that for us.
            Context.Response.StatusCode = 401;
            return Task.CompletedTask;
        }

        protected override Task HandleForbiddenAsync(AuthenticationProperties properties)
        {
            Context.Response.StatusCode = 403;
            return Task.CompletedTask;
        }
    }
```

Web.Config example:
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <system.webServer>
    <modules runAllManagedModulesForAllRequests="false">
      <remove name="WebDAVModule" />
    </modules>
    <security>
      <requestFiltering>
        <requestLimits maxUrl="8192" maxQueryString="4096" maxAllowedContentLength="512000000" />
      </requestFiltering>
    </security>
    <handlers>
      <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
    </handlers>
    <aspNetCore processPath="%LAUNCHER_PATH%" arguments="%LAUNCHER_ARGS%" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" hostingModel="outofprocess" />
  </system.webServer>
  
  <!--Prevent cache of index-->
  <location path="index.html">
    <system.webServer>
      <httpProtocol>
        <customHeaders>
          <add name="Cache-Control" value="no-cache" />
        </customHeaders>
      </httpProtocol>
    </system.webServer>
  </location>
  
  <!--Security-->
  <location path="api">
    <system.webServer>
      <security>
        <authentication>
          <windowsAuthentication enabled="false" />
        </authentication>
      </security>
    </system.webServer>
  </location>
  
  
  <location path="api/account/login/windows">
    <system.webServer>
      <security>
        <authentication>
          <windowsAuthentication enabled="true" />
        </authentication>
      </security>
    </system.webServer>
  </location>
  
  <location path="api/account/register/windows">
    <system.webServer>
      <security>
        <authentication>
          <windowsAuthentication enabled="true" />
        </authentication>
      </security>
    </system.webServer>
  </location>
  
</configuration>
```

Startup:
```
services.AddAuthentication()
.AddJwtBearer(options =>
{
	//blah
})
.AddWindowsAuthentication(options =>
{
	options.OnCreateClaims = (windowsPrincipal) =>
	{
		//blah

		return Task.FromResult(claims);
	};
})
```

On a last note, maybe we could also dedicate a documentation section for mixed authentication since SPAs are very popular these days?

Cheers!

**Comment by anurse** [comment]
Closing as per @Tratcher's answer. If you want to provide additional feedback on Negotiate that isn't directly related to the original post, feel free to create a new issue.

**Comment by Tratcher** [comment]
@treborgero looks like a lot of that WindowsAuthenticationHandler code was copied from IIS out-of-proc?
https://github.com/dotnet/aspnetcore/blob/18cb57aa3271e3e229ef9bdfe1e2b85500c53cb6/src/Servers/IIS/IISIntegration/src/IISMiddleware.cs#L160-L180

Looks like you've only added OnCreateClaims, but that could have been done with IClaimsTransformation as well.

**Comment by treborgero** [comment]
@Tratcher Yes sir! Here is the problem, the "IIS out-of-proc" logic doesn't play well with JWT (I can, for now, only confirm this for .NET Core 2.2) and you said it yourself in an [old issue](https://github.com/dotnet/aspnetcore/issues/4062). With that said, I went with my solution in the past but was hoping this new "**Microsoft.AspNetCore.Authentication.Negotiate**" package would solve the issue. My old solution, even though "hacky", allows for seamless integration with multi-factor authentication scenarios.

Most of my targeted audiences love the feeling of SPAs and have the need for intranet authentication locally and certificate/token authentication for other trusted networks. I proceed to authenticate users via windows or certificate authentication and then return a JWT back to my SPA client. To your second point, unfortunately the "**IClaimsTransformation** " approach is too generic and does not allow direct access to, for example, client's certificate info which would force developers to inject the HttpContext (e.g. httpContextAccessor.HttpContext.Request.HttpContext.Connection.ClientCertificate). On top of that the logic would contain conditionals for each authentication type. 

[Wouldn't it be nicer to have it in the "Startup" flow?](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/certauth?view=aspnetcore-3.1) 👼

**Comment by docmaster2** [comment]
всё отлично,я завёлся ©2020 DeviantArt.08.05.2020, 01:39, "Robert Orama" <notifications@github.com>:
@Tratcher Yes sir! Here is the problem, the "IIS out-of-proc" logic doesn't play well with JWT (I can, for now, only confirm this for .NET Core 2.2) and you said it yourself in an old issue. With that said, I went with my solution in the past but was hoping this new "Microsoft.AspNetCore.Authentication.Negotiate" package would solve the issue. My old solution, even though "hacky", allows for seamless integration with multi-factor authentication scenarios.
Most of my targeted audiences love the feeling of SPAs and have the need for intranet authentication locally and certificate/token authentication for other trusted networks. I proceed to authenticate users via windows or certificate authentication and then return a JWT back to my SPA client. To your second point, unfortunately the "IClaimsTransformation " approach is too generic and does not allow direct access to, for example, client's certificate info which would force developers to inject the HttpContext (e.g. httpContextAccessor.HttpContext.Request.HttpContext.Connection.ClientCertificate). On top of that the logic would contain conditionals for each authentication type.
Wouldn't it be nicer to have it in the "Startup" flow? 👼
